### PR TITLE
Fix Delete the main category of a product, product should be reassigned another main category

### LINKED ISSUE #9811: [BOOM-4293] Delete the main category of a product, product should be reassigned another main category
> This issue has been migrated from the Forge. Read the [original ticket here](http://forge.prestashop.com/browse/BOOM-4293).

- _**Reporter:**_ marion_francois
- _**Created at:**_ Mon, 13 Nov 2017 11:21:40 +0100

#### Describe the bug

When you associated a product to two categories and delete the default one with the option **If they have no other category, I want to associate them with the parent category.**. My product doesn't have a default category (NOK!).

#### Expected behavior

It should have the parent category as default category.

**TO SUM IT UP**: When I delete the main category of product A, then I am prompted a choice. If I choose that product A has to be re-associated with another category B, then category B must become the "default category" for product A.

#### Steps to reproduce

1. I associate my product to 2 categories: **Men** & **Home**
2. The default category for my products is **Men**
3. Delete the category **Men**
4. Choose to associate **the products without other categories** to the parent category, and keep them enabled.
5. My product is still associated to the Home category (good !) but doesn't have a default category (bad!).
       
**Screenshots**

https://drive.google.com/file/d/1BRyAAbhssGSG_Co8Y1GOZdYcNPYKMam1/view?usp=sharing

capture d'écran_192

#### Additional infos

- PrestaShop version: 1.7.7.3
- PHP version: 7.2

### PR DESCRIPTION
**Pull request by arouiadib**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | See discussion in BC Break PR https://github.com/PrestaShop/PrestaShop/pull/24223
| Type?             | bug fix 
| Category?         |BO 
| BC breaks?        |  no
| Deprecations?     |  no
| Fixed ticket?     | Fixes #9811 
| How to test?      | Steps to reproduce as in original ticket :<br>*  I associate my product to 2 categories: Men & Home<br>* The default category for my products is Men<br>* Delete the category Men<br>* Choose to associate the products without other categories to the parent category, and keep them enabled.<br>* My product is still associated to the Home category (good !) but doesn't have a default category (bad!).

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/24346)

### CODE REVIEW

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
Could you please describe what you changed and why? Regarding the table inside your PR description, we don't understand what you did :thinking:

**Comment by arouiadib** [comment]
Thanks @PierreRambaud for your time reviewing my PR.
I have edited the table describing it.

~~Note: There is also another discussion [on another PR (with same mods)](https://github.com/PrestaShop/PrestaShop/pull/24347) that I have made on 1.7.8.x~~

~~I believe it is a complex issue, so your time is invaluable to me so that I find/learn the best approach to fix it.~~

**Comment by arouiadib** [comment]
BTW @PierreRambaud , does `composer create-test-db` works these days? A behaviour test would be nice for this PR.

File: classes/Category.php [hunk 1]
```diff
@@ -2452,4 +2452,21 @@ public function isRootCategory(): bool
     {
         return 0 === (int) $this->id_parent;
     }
+
+    /**
+     * Get products of the category that will be deleted
+     */
+    public function getProductsBeforeDeletion(): array
+    {
+        $products = \Db::getInstance()->executeS('
```
Comment (Progi1984) [inline classes/Category.php]: Please move `Db` to use section.

File: classes/Category.php [hunk 2]
```diff
@@ -2043,6 +2043,23 @@ public function getProductsWs()
 		ORDER BY `position` ASC');
     }
 
+    /**
+     * Get products of the category that will be deleted
+     */
+    public function getProductsBeforeDeletion(): array
```
Comment (atomiix) [inline classes/Category.php]: I think the name of this function can be a bit more generic, like `getAllCategoryProducts()`

Comment (arouiadib) [reply classes/Category.php]: what about `getCategoryProducts`? 

Thanks for CR

Comment (Progi1984) [reply classes/Category.php]: @arouiadib Ok for this naming

File: src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php [hunk 3]
```diff
@@ -28,32 +28,23 @@
...
+     * Handle products update after category deletion.
      *
-     * @param int $parentCategoryId
-     * @param CategoryDeleteMode $mode
+     * @param array $categoryProducts products that belong to the category before deletion
+     * @param int $parentCategoryId parent category of the category to be deleted
+     * @param CategoryDeleteMode $mode deletion mode
      */
-    protected function handleProductsUpdate($parentCategoryId, CategoryDeleteMode $mode)
+    protected function handleProductsUpdate(array $categoryProducts, $parentCategoryId, CategoryDeleteMode $mode)
```
Comment (Progi1984) [inline src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php]: Could you move this parameter to the end and add a default value to `[]` for avoiding BC Break ?

**Review by Progi1984 — COMMENTED** [review]
Two small feedbacks :)

**Comment by arouiadib** [comment]
> Two small feedbacks :)

Great points here. Thanks.
I have suggested a solution to that, could you please see what I modified when you have time?

File: classes/Category.php [hunk 4]
```diff
@@ -2054,6 +2054,23 @@ public function getProductsWs()
...
+     */
+    public function getCategoryProducts()
+    {
+        return Db::getInstance()->executeS('
+        SELECT p.`id_product`
+        FROM `' . _DB_PREFIX_ . 'product` p
+        ' . Shop::addSqlAssociation('product', 'p') . '
+        LEFT JOIN `' . _DB_PREFIX_ . 'category_product` cp
+        ON cp.`id_product` = p.`id_product`
+        WHERE cp.`id_category` = ' . (int) $this->id);
```
Comment (Progi1984) [inline classes/Category.php]: Better : could you `sprintf` for the query ?

File: src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php [hunk 5]
```diff
@@ -28,32 +28,23 @@
...
+     * Handle products update after category deletion.
      *
-     * @param int $parentCategoryId
-     * @param CategoryDeleteMode $mode
+     * @param int $parentCategoryId parent category of the category to be deleted
+     * @param CategoryDeleteMode $mode deletion mode
+     * @param array $categoryProducts products of the category to be deleted
      */
-    protected function handleProductsUpdate($parentCategoryId, CategoryDeleteMode $mode)
+    protected function handleProductsUpdate($parentCategoryId, CategoryDeleteMode $mode, array $categoryProducts)
```
Comment (Progi1984) [inline src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php]:
```suggestion
    protected function handleProductsUpdate($parentCategoryId, CategoryDeleteMode $mode, array $categoryProducts = [])
```

**Review by Progi1984 — COMMENTED** [review]
As you add behavior in two handlers, could you add behat tests ? After that, it will be an approve for me.

**Comment by arouiadib** [comment]
> As you add behavior in two handlers, could you add behat tests ? After that, it will be an approve for me.

Totally agree with you. I have already tried here: https://github.com/PrestaShop/PrestaShop/pull/24668 but it wasn't obvious for me to complete them because of my lack of experience there.

I will retry today and any future help will be much appreciated.

**Comment by arouiadib** [comment]
@Progi1984 It seems I finally managed to create my first behavior test (ever).

They are green here and [ the same tests in develop are red.](https://github.com/PrestaShop/PrestaShop/pull/24668)

File: classes/Category.php (hunk 4 above)
Comment (arouiadib) [reply classes/Category.php]: Fixed. (But I feel I should make it better by using category Repository and rely on ORM, maybe inject the service `category.repository` somewhere :thinking: .)

File: tests/Integration/Behaviour/Features/Context/Domain/CategoryFeatureContext.php [hunk 6]
```diff
@@ -517,6 +519,39 @@ public function assertCategoryExistsByName(string $categoryReference, string $ca
...
+        $addProductFeatureContext->addProduct($arg1, $table);
+    }
+
+    /**
+     * @When I set product :productReference to following categories:
+     *
+     * @param string $productReference
+     * @param TableNode $table
+     */
+    public function iAssignProductProductToFollowingCategories(string $productReference, TableNode $table)
```
Comment (Progi1984) [inline tests/Integration/Behaviour/Features/Context/Domain/CategoryFeatureContext.php]:
```suggestion
    public function iAssignProductProductToFollowingCategories(string $productReference, TableNode $table): void
```

File: tests/Integration/Behaviour/Features/Context/Domain/CategoryFeatureContext.php [hunk 7]
```diff
@@ -517,6 +519,39 @@ public function assertCategoryExistsByName(string $categoryReference, string $ca
...
+        $updateCategoriesContext->assignToCategoriesIncludingNonExistingOnes($productReference, $table);
+    }
+
+    /**
+     * @Then the product :productReference should be assigned to following categories:
+     *
+     * @param string $productReference
+     * @param TableNode $table
+     */
+    public function assertProductCategories(string $productReference, TableNode $table)
```
Comment (Progi1984) [inline tests/Integration/Behaviour/Features/Context/Domain/CategoryFeatureContext.php]:
```suggestion
    public function assertProductCategories(string $productReference, TableNode $table): void
```

**Review by Progi1984 — COMMENTED** [review]
A small type hint and it's approved ;)

File: tests/Integration/Behaviour/Features/Context/Domain/CategoryFeatureContext.php [hunk 8]
```diff
@@ -517,6 +519,39 @@ public function assertCategoryExistsByName(string $categoryReference, string $ca
         $this->getSharedStorage()->set($categoryReference, (int) $foundCategory['id_category']);
     }
 
+    /**
+     * @When I create product :arg1 with following information:
+     */
+    public function iAddProductWithFollowingInformation($arg1, TableNode $table)
```
Comment (Progi1984) [inline tests/Integration/Behaviour/Features/Context/Domain/CategoryFeatureContext.php]:
```suggestion
    public function iAddProductWithFollowingInformation(string $arg1, TableNode $table): void
```

**Review by Progi1984 — COMMENTED** [review]
Thank you for your contribution. I forgot a suggestion on type hinting.

**Comment by arouiadib** [comment]
> Thank you for your contribution. I forgot a suggestion on type hinting.

It is my mistake to not apply your CR on whole PR. Thanks a lot @Progi1984

**Review by Progi1984 — DISMISSED** [review]

**Comment by atomiix** [comment]
Hey @arouiadib, can you fix the conflicting file? :)

**Comment by atomiix** [comment]
@arouiadib It seems that you have a merge commit in your git history, can you rebase instead of pulling so you won't have a merge commit? Thanks :)

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
Could you please rebase and squash your commits, seems commits have no real meaning :thinking: 
image

**Comment by arouiadib** [comment]
@atomiix , @PierreRambaud Sure guys, thanks

**Comment by PierreRambaud** [comment]
Hi @arouiadib ,

Since we had no news from you for more than 30 days, I'll close this pull request. Feel free to reopen or open another one if you think it's still relevant.

Thanks!

**Pull request closed without merging**